### PR TITLE
[Federation] Install the required firewall rule in the federated Ingress test.

### PR DESCRIPTION
**Pull request by madhusudancs**
```release-note
Federated Ingress, released as alpha in v1.4, sometimes causes the health checks corresponding to the backends in the underlying clusters to flap. Please use the workaround described in the Federated Ingress documentation - http://kubernetes.io/docs/user-guide/federation/federated-ingress/#known-issue. This will be fixed in v1.6.
```

This works around issue #36327.

Also, this PR doesn't fix all the Federated Ingress problems. It resolves the infrastructure issues. The real test which is in the loop here - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/federated-ingress.go#L420 doesn't even run because len(Rules) is 0. I want to get this PR in to get our tests to a stable state while I debug the tests and the controller.

cc @kubernetes/sig-cluster-federation @nikhiljindal

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37571)

### CODE REVIEW

**Comment by dims** [comment]
@k8s-bot cvm gke e2e test this

**Comment by dims** [comment]
@k8s-bot gci gke e2e test this

**Comment by dims** [comment]
@k8s-bot verify test this

**Comment by dims** [comment]
@madhusudancs looks like the bazel stuff needs to be updated. ( cc @matchstick )

**Comment by dims** [comment]
@madhusudancs Move the milestone please if appropriate?

**Comment by madhusudancs** [comment]
@dims no. This has to be cherrypicked to one of the v1.5 patch releases.

File: test/test_owners.csv [hunk 1]
```diff
@@ -63,7 +63,7 @@ Density create a batch of pods latency/resource should be within limit when crea
 Density create a batch of pods with higher API QPS latency/resource should be within limit when create * pods with * interval (QPS *),jlowdermilk,1
 Density create a sequence of pods latency/resource should be within limit when create * pods with * background pods,wojtek-t,1
 Density should allow running maximum capacity pods on nodes,smarterclayton,1
-Density should allow starting * pods per node using *,derekwaynecarr,0
+Density should allow starting * pods per node using * with * secrets,madhusudancs,0
```
Comment (madhusudancs) [inline test/test_owners.csv]: Erm, this doesn't make any sense.

@kubernetes/test-infra-maintainers any idea?

**Comment by madhusudancs** [comment]
@k8s-bot federation gce e2e test this

**Comment by madhusudancs** [comment]
@k8s-bot federation gce e2e test this

File: test/e2e/federated-ingress.go [hunk 2]
```diff
@@ -484,3 +513,119 @@ func getFederatedIngressAddress(client *fedclientset.Clientset, ns, name string)
 	}
 	return addresses, nil
 }
+
+// getNodeNames returns the names of nodes in a cluster.
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: comment is out of sync with code

File: test/e2e/federated-ingress.go [hunk 3]
```diff
@@ -484,3 +513,119 @@ func getFederatedIngressAddress(client *fedclientset.Clientset, ns, name string)
...
+func getZoneNodesMap(clientset *kubeclientset.Clientset) (map[string][]string, error) {
+	zoneNodes := make(map[string][]string)
+	nodes, err := clientset.Core().Nodes().List(v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	// Only consider schedulable nodes, because we won't
+	// have backends in non-schedulable nodes and don't
+	// want to open port to those nodes.
+	// Also, master node is unschedulable and contains
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: We should not assume here that master node is unschedulable. Using Spec.Unschedulable should be enough

Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: Thats what you seem to be doing in code. So just update this comment?

File: test/e2e/federated-ingress.go [hunk 4]
```diff
@@ -484,3 +513,119 @@ func getFederatedIngressAddress(client *fedclientset.Clientset, ns, name string)
...
+		if !ok {
+			nodes = make([]string, 0)
+		}
+		nodes = append(nodes, n.Name)
+		zoneNodes[zone] = nodes
+	}
+	return zoneNodes, nil
+}
+
+func updateMap(merged, src map[string][]string) {
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: Is mergeMap a better name?

File: test/e2e/federated-ingress.go [hunk 5]
```diff
@@ -484,3 +513,119 @@ func getFederatedIngressAddress(client *fedclientset.Clientset, ns, name string)
...
+// https://github.com/kubernetes/kubernetes/issues/37306
+// is implemented.
+func createFederatedIngressFirewallRules(ns, ingressName string, clusters map[string]*cluster, svc *v1.Service) (map[string]string, error) {
+	mergedZoneNodes := make(map[string][]string)
+	for clusterName, cluster := range clusters {
+		zoneNodes, err := getZoneNodesMap(cluster.Clientset)
+		if err != nil {
+			return nil, err
+		}
+		framework.Logf("nodes to zones map for cluster %q: %#v", clusterName, mergedZoneNodes)
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: Make them glog.V(2)?

File: test/e2e/federated-ingress.go [hunk 6]
```diff
@@ -484,3 +513,119 @@ func getFederatedIngressAddress(client *fedclientset.Clientset, ns, name string)
...
+			return fwNames, err
+		}
+		fwNames[zone] = fwName
+	}
+	return fwNames, nil
+}
+
+func createFederatedIngressFirewallRulesOrFail(ns, ingressName string, clusters map[string]*cluster, svc *v1.Service) map[string]string {
+	fwNames, err := createFederatedIngressFirewallRules(ns, ingressName, clusters, svc)
+	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Error creating firewall rule for Ingress %q in %q", ingressName, ns))
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: Also include the original err in error message

File: test/e2e/federated-ingress.go [hunk 7]
```diff
@@ -484,3 +513,119 @@ func getFederatedIngressAddress(client *fedclientset.Clientset, ns, name string)
...
+	var errs []string
+	for zone, fwName := range fwNames {
+		By(fmt.Sprintf("Deleting firewall rule %q", fwName))
+		provider, ok := framework.TestContext.CloudConfig.FederationProviders[zone]
+		if !ok {
+			return fmt.Errorf("no cloud provider for zone %q", zone)
+		}
+		gce := provider.(*gceprovider.GCECloud)
+		err := gce.DeleteFirewall(fwName)
+		// Collect the erro and continue
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: erro -> error

File: test/e2e/federated-ingress.go [hunk 8]
```diff
@@ -43,9 +43,10 @@ import (
 
 const (
 	MaxRetriesOnFederatedApiserver = 3
-	FederatedIngressTimeout        = 120 * time.Second
+	FederatedIngressTimeout        = 10 * time.Minute
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: This is a huge increase :)

File: test/e2e/federated-ingress.go [hunk 9]
```diff
@@ -464,17 +515,34 @@ func (j *federationTestJig) waitForFederatedIngress() {
...
-			framework.Logf("Testing route %v host %v with simple GET", route, rules.Host)
-			framework.ExpectNoError(pollURL(route, rules.Host, lbPollTimeout, lbPollInterval, timeoutClient, false))
-		}
+
+	client := &http.Client{
+		// This is mostly `http.DefaultTransport` except for the
+		// `TLSClientConfig`. `http.DefaultTransport` cannot be
+		// shallow copied due to the unexported `sync.Mutex` in
+		// the struct.
+		Transport: &http.Transport{
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: There should be something better
Does using SetTransportDefaults work? Example usage: https://github.com/kubernetes/kubernetes/blob/8fd414537b5143ab039cb910590237cabf4af783/pkg/probe/http/http.go#L35

File: test/e2e/federated-ingress.go [hunk 10]
```diff
@@ -58,8 +61,56 @@ const (
...
+	FederatedIngressTLSCrt = `-----BEGIN CERTIFICATE-----
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: Atleast put them in a separate file?

File: test/e2e/federated-ingress.go [hunk 11]
```diff
@@ -58,8 +61,56 @@ const (
 	FederatedIngressGCLBSrcRange = "130.211.0.0/22"
 
 	// TLS Certificate and Key for the ingress resource
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: Add instructions on how you generated them to make it easier for the guy who wants to change these

File: test/e2e/federated-ingress.go (hunk 7 above)
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: ignore I see you fixed it in next commit

File: test/e2e/federated-ingress.go (hunk 6 above)
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: Ignore I see you fixed it in next commit

File: test/e2e/federated-ingress.go (hunk 10 above)
Comment (madhusudancs) [inline test/e2e/federated-ingress.go]: Needs file reading and the associated error handling. Plus you need to ensure that the right files get copied to the test tarballs. It is just too painful. I would just leave it here. Also, it is a commonly used test pattern. Ex: https://github.com/golang/go/blob/master/src/crypto/tls/tls_test.go#L26

File: test/e2e/federated-ingress.go (hunk 11 above)
Comment (madhusudancs) [inline test/e2e/federated-ingress.go]: It doesn't really matter. Any method works, but added it that in the comment.

File: test/e2e/federated-ingress.go (hunk 9 above)
Comment (madhusudancs) [inline test/e2e/federated-ingress.go]: Good idea. Thanks for pointing it out. Did not know its existence.

File: test/e2e/e2e.go [hunk 12]
```diff
@@ -76,6 +78,20 @@ func setupProviderConfig() error {
 			return fmt.Errorf("Error building GCE/GKE provider: %v", err)
 		}
 
+		// TODO(madhusudancs): Remove this once
```
Comment (madhusudancs) [inline test/e2e/e2e.go]: That comment is in https://github.com/kubernetes/kubernetes/pull/37571/commits/3806de3752ddd3a86d6ed6740b71613bce4208e7#diff-e00287b5331f6d39959423afadb6b368R134. But also added a comment here.

File: test/e2e/federated-ingress.go (hunk 2 above)
Comment (madhusudancs) [inline test/e2e/federated-ingress.go]: Updated! Thanks!

File: test/e2e/federated-ingress.go (hunk 3 above)
Comment (madhusudancs) [inline test/e2e/federated-ingress.go]: It is an additional comment. Like a caveat.

File: test/e2e/federated-ingress.go (hunk 4 above)
Comment (madhusudancs) [inline test/e2e/federated-ingress.go]: Changed.

File: test/e2e/federated-ingress.go (hunk 5 above)
Comment (madhusudancs) [inline test/e2e/federated-ingress.go]: Why? We rarely use glog in e2e test code. It is incredibly useful while debugging in a non-interactive run, meaning jenkins.

File: test/e2e/federated-ingress.go (hunk 8 above)
Comment (madhusudancs) [inline test/e2e/federated-ingress.go]: Yeah. But Ingresses take a long time to propagate.

**Review by madhusudancs — COMMENTED** [review]
@nikhiljindal addressed review comments. PTAL.

**Comment by quinton-hoole** [comment]
Can we please split this PR up into multiple separate PR's?  In it's current form, this PR does 5 largely independent things, as far as I can tell:

1. Add firewall rules
2. Change test owners
3. Change test artifact names
4. Change timeouts
5. Change endpoints 

It seems that each should be submitted as a separate PR.

Also, as discussed in #37306, I'm not sure that we necessarily want 1 (as it will ultimately need to be reverted once #37306 has been fixed.  Either way, reverting it will be easier if it's in a separate PR.

**Comment by madhusudancs** [comment]
@quinton-hoole So the serious problems mentioned in https://github.com/kubernetes/kubernetes/issues/37306#issuecomment-279498697 is about splitting this PR?

> It seems that each should be submitted as a separate PR.

Changing test owners is a requirement to get the PR in. They are there to make presubmits happy. This PR won't go in otherwise.

> Also, as discussed in #37306, I'm not sure that we necessarily want 1 (as it will ultimately need to be reverted once #37306 has been fixed). Either way, reverting it will be easier if it's in a separate PR.

We don't revert PRs. We revert git commits. This is the primary reason why these changes you listed are in separate git commits. You can atomically revert 1 when #37306 is fixed.

**Comment by madhusudancs** [comment]
> We don't revert PRs. We revert git commits. This is the primary reason why these changes you listed are in separate git commits. You can atomically revert 1 when #37306 is fixed.

Oops, I meant. We don't **have to** revert PRs. We **can** revert git commits.

**Comment by quinton-hoole** [comment]
Fair enough, as long as we don't squash the commits before merging (which is what we usually do).
Also, I think that we usually revert PR's, not individual commits?  I don't feel terribly strongly either way, as long as it's easy to revert each one after the fact if required.

As for the "serious problems", at first glance I thought that the firewall creation logic was in the federated ingress controller, as proposed in #37306.  But I see it's in the test - my apologies for misinterpreting it initially.

**Comment by madhusudancs** [comment]
@quinton-hoole thanks for the clarifications.

**Comment by csbell** [comment]
Closing here since the issue is addressed in #41942. Still requires a matching release of GLBC/Ingress to be visible in Jenkins.

**Comment by nikhiljindal** [comment]
Is there an issue tracking release of GLBC? Will it be part of 1.6?

**Comment by csbell** [comment]
https://github.com/kubernetes/ingress/issues/381

**Pull request closed without merging**